File: server/routes/appointments.js
const { Router } = require("express");
const { v4: uuid } = require("uuid");
const db = require("../db");
const { authenticate } = require("../auth");
const { availableSlots, slotMinutes } = require("../scheduling");

const router = Router();

router.get("/reasons", (req, res) => {
  const reasons = db.prepare("SELECT * FROM reasons_for_visit").all();
  res.json(reasons.map((r) => ({
    id: r.id,
    label: r.label,
    is_urgent_default: !!r.is_urgent_default,
  })));
});

router.post("/appointments", authenticate, (req, res) => {
  const { doctor_id, reason_id, starts_at, mode, symptoms } = req.body;

  const doctor = db.prepare("SELECT id FROM doctors WHERE id = ?").get(doctor_id);
  if (!doctor) return res.status(404).json({ detail: "Doctor not found" });

  // Validate slot is open
  const day = starts_at.slice(0, 10);
  const openSlots = availableSlots(doctor_id, day);
  const openStarts = new Set(openSlots.map((s) => s.starts_at));
  if (!openStarts.has(starts_at)) {
    return res.status(409).json({ detail: "Slot is not available" });
  }

  // Compute end time
  const d = new Date(starts_at);
  const weekday = (d.getDay() + 6) % 7;
  const mins = slotMinutes(doctor_id, weekday);
  const endDate = new Date(d.getTime() + mins * 60000);
  const ends_at = endDate.toISOString().replace(/\.\d{3}Z$/, "").replace("Z", "");

  const id = uuid();
  const visitMode = mode || "in_clinic";
  let telehealthUrl = null;
  if (visitMode === "telehealth") {
    telehealthUrl = `https://video.orthocare.example/room/${id}`;
  }

  try {
    db.prepare(
      `INSERT INTO appointments (id, patient_id, doctor_id, reason_id, mode, status, starts_at, ends_at, symptoms, telehealth_url)
       VALUES (?, ?, ?, ?, ?, 'confirmed', ?, ?, ?, ?)`
    ).run(id, req.user.id, doctor_id, reason_id || null, visitMode, starts_at, ends_at, symptoms || null, telehealthUrl);
  } catch (err) {
    if (err.message.includes("UNIQUE")) {
      return res.status(409).json({ detail: "Slot just got taken" });
    }
    throw err;
  }

  const appt = db.prepare("SELECT * FROM appointments WHERE id = ?").get(id);
  console.log(`CONFIRMATION -> appointment ${appt.id} at ${appt.starts_at}`);
  res.status(201).json(appt);
});

router.get("/appointments/me", authenticate, (req, res) => {
  let appts;
  if (req.user.role === "patient") {
    appts = db.prepare("SELECT * FROM appointments WHERE patient_id = ? ORDER BY starts_at DESC").all(req.user.id);
  } else if (req.user.role === "doctor") {
    const doc = db.prepare("SELECT id FROM doctors WHERE user_id = ?").get(req.user.id);
    appts = doc
      ? db.prepare("SELECT * FROM appointments WHERE doctor_id = ? ORDER BY starts_at DESC").all(doc.id)
      : [];
  } else {
    appts = db.prepare("SELECT * FROM appointments ORDER BY starts_at DESC").all();
  }
  res.json(appts);
});

router.post("/appointments/:id/cancel", authenticate, (req, res) => {
  const appt = db.prepare("SELECT * FROM appointments WHERE id = ?").get(req.params.id);
  if (!appt) return res.status(404).json({ detail: "Appointment not found" });

  const isOwner = appt.patient_id === req.user.id;
  const isStaff = ["doctor", "admin"].includes(req.user.role);
  if (!isOwner && !isStaff) return res.status(403).json({ detail: "Not allowed" });

  db.prepare("UPDATE appointments SET status = 'cancelled' WHERE id = ?").run(appt.id);
  const updated = db.prepare("SELECT * FROM appointments WHERE id = ?").get(appt.id);
  res.json(updated);
});

module.exports = router;

File: backend/app/routers/appointments.py
from datetime import timedelta

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.exc import IntegrityError
from sqlalchemy.orm import Session

from ..database import get_db
from ..models import (
    Appointment, Doctor, AvailabilityRule, ReasonForVisit,
    AppointmentStatus, User, Role, VisitMode,
)
from ..schemas import AppointmentCreate, AppointmentOut, ReasonOut
from ..security import get_current_user, require_roles
from ..services.locks import acquire_slot, release_slot
from ..services.scheduling import available_slots
from ..services import notifications

router = APIRouter(prefix="/api", tags=["appointments"])


@router.get("/reasons", response_model=list[ReasonOut])
def list_reasons(db: Session = Depends(get_db)):
    return db.query(ReasonForVisit).all()


def _slot_minutes(db: Session, doctor_id: str, weekday: int) -> int:
    rule = (db.query(AvailabilityRule)
            .filter(AvailabilityRule.doctor_id == doctor_id,
                    AvailabilityRule.weekday == weekday).first())
    return rule.slot_minutes if rule else 30


@router.post("/appointments", response_model=AppointmentOut, status_code=201)
def book(payload: AppointmentCreate, user: User = Depends(get_current_user),
         db: Session = Depends(get_db)):
    doctor = db.query(Doctor).filter(Doctor.id == payload.doctor_id).first()
    if not doctor:
        raise HTTPException(status_code=404, detail="Doctor not found")

    # Validate the requested slot is genuinely open for that day.
    day = payload.starts_at.date()
    open_starts = {s["starts_at"] for s in available_slots(db, doctor.id, day)}
    if payload.starts_at not in open_starts:
        raise HTTPException(status_code=409, detail="Slot is not available")

    iso = payload.starts_at.isoformat()
    if not acquire_slot(doctor.id, iso):
        raise HTTPException(status_code=409, detail="Slot is being booked by someone else")

    try:
        minutes = _slot_minutes(db, doctor.id, payload.starts_at.weekday())
        appt = Appointment(
            patient_id=user.id,
            doctor_id=doctor.id,
            reason_id=payload.reason_id,
            mode=payload.mode,
            status=AppointmentStatus.confirmed,
            starts_at=payload.starts_at,
            ends_at=payload.starts_at + timedelta(minutes=minutes),
            symptoms=payload.symptoms,
        )
        if payload.mode == VisitMode.telehealth:
            # Placeholder; replace with Twilio Video / Zoom room creation.
            appt.telehealth_url = f"https://video.orthocare.example/room/{appt.id}"
        db.add(appt)
        db.commit()
        db.refresh(appt)
    except IntegrityError:
        db.rollback()
        release_slot(doctor.id, iso)
        raise HTTPException(status_code=409, detail="Slot just got taken")

    notifications.send_appointment_confirmation(appt)
    notifications.schedule_reminder(appt)  # 24h-before reminder (enqueue)
    return appt


@router.get("/appointments/me", response_model=list[AppointmentOut])
def my_appointments(user: User = Depends(get_current_user), db: Session = Depends(get_db)):
    q = db.query(Appointment)
    if user.role == Role.patient:
        q = q.filter(Appointment.patient_id == user.id)
    elif user.role == Role.doctor and user.doctor_profile:
        q = q.filter(Appointment.doctor_id == user.doctor_profile.id)
    return q.order_by(Appointment.starts_at.desc()).all()


@router.post("/appointments/{appt_id}/cancel", response_model=AppointmentOut)
def cancel(appt_id: str, user: User = Depends(get_current_user),
           db: Session = Depends(get_db)):
    appt = db.query(Appointment).filter(Appointment.id == appt_id).first()
    if not appt:
        raise HTTPException(status_code=404, detail="Appointment not found")
    is_owner = appt.patient_id == user.id
    is_staff = user.role in (Role.doctor, Role.admin)
    if not (is_owner or is_staff):
        raise HTTPException(status_code=403, detail="Not allowed")
    appt.status = AppointmentStatus.cancelled
    db.commit()
    release_slot(appt.doctor_id, appt.starts_at.isoformat())
    db.refresh(appt)
    return appt
